### ISSUE TITLE
`ray.get` hangs on `ObjectRef`s passed out-of-band

### ISSUE THREAD

**Issue by mitar** [ray-project/ray#9578]
### What is the problem?

Tested on Ray 0.8.6.

The problem is that it seems I cannot store and retrieve arbitrary objects outside of tasks/actors anymore. Documentation for `ray.put` and `ray.get` seems to reflect that this is still possible, but it is not really. See the reproduction.

For us this is now a major blocker. We used Ray (and before Plasma directly) to pass objects between user interface and workers. On user interface user would upload data and we would call `ray.put` and then over a custom API send the hex object ID to the main server which would then call Ray task. Now, that task simply hangs, `ray.get` never to return.

I understand that there is now fancy reference counting and things like that, but simple low-level `ray.get` and `ray.put` should still be possible. I think semantics should be that when `ray.get` calls an object ID not otherwise known to be used by this task, the task blocks, object is send to the task, and then task unblocks. This is how it was before. Even more, before during the time the task was blocked, the worker was in fact freed and other tasks could run there.

So, why does `ray.get` blocks? Why is not object eventually retrieved?

I would also be OK if those objects are simply eventually removed from memory, when there is pressure on storage. And once removed, if you call `ray.get` then, an exception is thrown. That is all OK. (While reference counting can help not remove something when it is known that it is still needed.) But that it just blocks without any error, instead of pulling the object to the worker, this is really strange and makes it useless and prevents out-of-band `ray.put` and `ray.get`.

### Reproduction

```python
import ray
from ray import utils

ray.init()

@ray.remote
def foo(object_id_hex):
    print(ray.get(utils.binary_to_object_id(utils.hex_to_binary(object_id_hex))))

object_id = ray.put(42)
ray.get(foo.remote(object_id.hex()))
print("finished")
```

`finished` is newer printed.

**Comment by mitar** [comment]
I think the behavior should be:;

```
object_id = utils.binary_to_object_id(utils.hex_to_binary(object_id_hex))
ray.get(object_id)
```

When `ray.get` is called on `object_id`, the referenced object should be fetched and returned. At the same time the object should be pinned until a reference to used `object_id` exists. So this is similar to how it works for `ray.put`.

**Comment by stephanie-wang** [comment]
It's no longer recommended to use the binary string of the ObjectRef directly. You can use `ray.cloudpickle` instead to serialize objects out of band. Note that we don't explicitly support out-of-band object refs, so it isn't guaranteed to work well with the current ref counting protocol.

```
import ray

ray.init()

@ray.remote
def foo(serialized_object_ref):
    print(ray.get(ray.cloudpickle.loads(serialized_object_ref)))

object_ref = ray.put(42)
ray.get(foo.remote(ray.cloudpickle.dumps(object_ref)))
print("finished")
```

**Comment by mitar** [comment]
Sure, but that makes very large objects. I want to not have to have that performance hit.

So can this be made into a feature request to do support this?

**Comment by stephanie-wang** [comment]
The serialized ref contains the address of the worker that "owns" the object, and eventually this will also be needed to also find the location of the object. So we can't make the serialized refs any smaller than that. It's only 184 bytes total, though, compared to 20 bytes for the raw ID string. Sorry if this is a blocker for you, but it's unlikely that we'll change the serialization format given that no one else has mentioned the serialized object size as a performance bottleneck yet and it's quite ingrained in the architecture.

Other than the serialization overhead, is there any behavior missing from the current system that your use case requires?

**Comment by mitar** [comment]
Oh, sorry. I didn't see that. :-) So this is pickle of the object ref. Yes, I can do that. :-)

**Comment by mitar** [comment]
Does it have to be cloudpickle or does the regular pickle works? Does it have to be `ray.cloudpickle`?

**Comment by stephanie-wang** [comment]
`ray.cloudpickle` is needed because it registers a custom serializer/deserializer that dumps/loads the owner's address along with the object's ID string. I believe if you use regular pickle, it will just dump the ID string, and then you'll have the same issue as before.

**Comment by mitar** [comment]
> Note that we don't explicitly support out-of-band object refs, so it isn't guaranteed to work well with the current ref counting protocol.

Also, about this. Could there be some guarantees here? Like, I mention above. That reference counting would also count users of the reference, not just owners?

And how long lasting are such pickled object references? What if the original worker is not available anymore, but the object is still somewhere in the cluster? Can such object reference still resolve?

Also, could you then throw an exception if one tries to use object id like I did? Instead of just hanging?

**Comment by stephanie-wang** [comment]
What I mean by "not guaranteed to work" is that it's possible the objects will get cleaned up while there's still a serialized copy of the object somewhere. But you will be guaranteed to receive an error if you call `ray.get` from an out-of-band copy and the object is already gone or the owner is dead.

Right now, objects will get eventually cleaned up after their in-band ref count goes to 0. Unfortunately, it's impossible to keep track of users of the serialized ref since that's just a string that can be passed around and stored in arbitrary ways. This is why the ref counting protocol is based on `ObjectRef`s, which are Python objects whose lifetimes we can track.

Pickled object refs will last as long as the original worker process is still alive. After the owner is dead, you will not hang, but you may not be able to get a copy of the object anymore. I think it may be a good idea to support a feature where the ref can be long-lived, past the original owner's lifetime, but the user would then be responsible for memory management (deleting the object ref once it is no longer needed). You can file a separate feature request for that if you're interested, to gauge community interest. We haven't prioritized it since there are usually workarounds.

Yes, we could extend the `ObjectRef` class instead, but there hasn't really been a need yet.

**Comment by mitar** [comment]
OK, I am a bit unsure what is `ObjectRef` and its relation to `ObjectId`? I assume the same thing.

> it's possible the objects will get cleaned up while there's still a serialized copy of the object somewhere

I think this is how it was before? Or before plasma store centrally tracked all copies and so it was able to give you one while there was at least one copy somewhere in the cluster? And this is not possible anymore? You can get it only from original owner? Do regular `ObjectRef`s know how to obtain object from another place in the cluster if owner dies? Or does the owner first have to be recreated?

> Unfortunately, it's impossible to keep track of users of the serialized ref since that's just a string that can be passed around and stored in arbitrary ways. This is why the ref counting protocol is based on ObjectRefs, which are Python objects whose lifetimes we can track.

Sure, but once the "serialized ref" is restored back to `ObjectRef`, you can start counting it again, no? This means that I could create `ObjectRef`, serialize it, pass it out of band, deseriaize it, and owner could then forget it, because the new destination would now hold the reference?

> past the original owner's lifetime, but the user would then be responsible for memory management (deleting the object ref once it is no longer needed).

I mean, if you would count all `ObjectRef` objects in any Ray Python processes, owner or consumer, this would not be a problem and user would not have to manually do memory management? So I do not see why are you counting references just for owners, when you could also count whoever creates an `ObjectRef`?

A side question: you are counting just instances of `ObjectRef`s in Python processes, no? Why not also objects the `ObjectRef`s are pointing to? So I would assume that if either of those two are in process memory, reference count should be non-zero and other `ObjectRef`s should be able to retrieve the object?

Can I then suggest the following:
* You get rid of serialization to hex (so no more method `.hex()` and so on). It seems it does not provide enough state.
  * That also solves the problem that I think `ray.get` should throw an error if a bad (partial state) `ObjectRef` is provided instead of current hanging. If the above is not done, then at least throw an error.
  * Or, you could make `.hex()` return full state of any `ObjectRef`.
* `ObjectRef` class is extended so that it works with regular pickle and cloudpickle, I really do not see a reason why this could not be made so, one then does not have to think about that.
   * If this is not done, `ray.get` should raise an error if regular pickle is used, instead of just hanging around.
* When `ObjectRef` is constructed from pickle, it should register itself as ref count. Now is Python object and can be counted.
* Reference counting should count both the `ObjectRef`s and objects themselves, where if either of them are still available in memory, the reference count should be non zero and object should be available to others.

Can this issue become the above list or should I open new issues for each?

**Comment by stephanie-wang** [comment]
> OK, I am a bit unsure what is `ObjectRef` and its relation to `ObjectId`? I assume the same thing.
They're the same thing. We're switching over to `ObjectRef` in the next Ray release.

> 
> > it's possible the objects will get cleaned up while there's still a serialized copy of the object somewhere
> 
> I think this is how it was before? Or before plasma store centrally tracked all copies and so it was able to give you one while there was at least one copy somewhere in the cluster? And this is not possible anymore? You can get it only from original owner? Do regular `ObjectRef`s know how to obtain object from another place in the cluster if owner dies? Or does the owner first have to be recreated?

I'm not sure what you mean about "before", but in future versions of Ray, you will only be able to get the object while the original owner is alive.

> 
> > Unfortunately, it's impossible to keep track of users of the serialized ref since that's just a string that can be passed around and stored in arbitrary ways. This is why the ref counting protocol is based on ObjectRefs, which are Python objects whose lifetimes we can track.
> 
> Sure, but once the "serialized ref" is restored back to `ObjectRef`, you can start counting it again, no? This means that I could create `ObjectRef`, serialize it, pass it out of band, deseriaize it, and owner could then forget it, because the new destination would now hold the reference?

Yes, we could track the deserialized `ObjectRef`s, but I don't see how this would gain you much. There's still a chance the object could go out of scope while it is being passed out of band.

> 
> > past the original owner's lifetime, but the user would then be responsible for memory management (deleting the object ref once it is no longer needed).
> 
> I mean, if you would count all `ObjectRef` objects in any Ray Python processes, owner or consumer, this would not be a problem and user would not have to manually do memory management? So I do not see why are you counting references just for owners, when you could also count whoever creates an `ObjectRef`?

Again, it's impossible to do ref counting for serialized copies of an `ObjectRef`.

> 
> A side question: you are counting just instances of `ObjectRef`s in Python processes, no? Why not also objects the `ObjectRef`s are pointing to? So I would assume that if either of those two are in process memory, reference count should be non-zero and other `ObjectRef`s should be able to retrieve the object?
> 
> Can I then suggest the following:
> 
>     * You get rid of serialization to hex (so no more method `.hex()` and so on). It seems it does not provide enough state.
>       
>       * That also solves the problem that I think `ray.get` should throw an error if a bad (partial state) `ObjectRef` is provided instead of current hanging. If the above is not done, then at least throw an error.
>       * Or, you could make `.hex()` return full state of any `ObjectRef`.

Sure, that makes sense. I think it's just oversight that the invalid `ObjectRef` can be used at all right now.

> 
>     * `ObjectRef` class is extended so that it works with regular pickle and cloudpickle, I really do not see a reason why this could not be made so, one then does not have to think about that.

We can do it; it's just not high priority since there's a reasonable workaround. I'd suggest opening a separate feature request for this since it's different from the above.

>       
>       * If this is not done, `ray.get` should raise an error if regular pickle is used, instead of just hanging around.
> 
>     * When `ObjectRef` is constructed from pickle, it should register itself as ref count. Now is Python object and can be counted.
> 
>     * Reference counting should count both the `ObjectRef`s and objects themselves, where if either of them are still available in memory, the reference count should be non zero and object should be available to others.

We already do this, just not for `ObjectRef`s that have been passed out-of-band. Again, not a high priority since there's a reasonable workaround and would suggest opening a separate feature request for it.

> 
> 
> Can this issue become the above list or should I open new issues for each?

**Comment by mitar** [comment]
> I'm not sure what you mean about "before"

In version 0.7.5 and earlier, where you still used Arrow and Arrow Plasma store. :-) I could there store things into Plasma store directly and pass serialized ObjectId as hex around and things just worked. Until object was removed from the store because of memory pressure. I am not sure if then there was enough to have only one copy of the object somewhere in the cluster (in local Plasma store on a node).

> Yes, we could track the deserialized ObjectRefs, but I don't see how this would gain you much. There's still a chance the object could go out of scope while it is being passed out of band.

But once the object is recreated in consumer, the original owner can go away, and third consumer can obtain the object. Or not?

So does ownership transfer in Ray if ObjectRefs are send in-band? Or is ownership shared?

> it's impossible to do ref counting for serialized copies of an ObjectRef.

I am not saying that. But you can count deserialized copies of ObjectRef. Are you doing that?

> Again, not a high priority since there's a reasonable workaround and would suggest opening a separate feature request for it.

Which workaround you have in mind here? Cloudpickle? Using cloudpickle addresses this?

**Comment by stephanie-wang** [comment]
> > Yes, we could track the deserialized ObjectRefs, but I don't see how this would gain you much. There's still a chance the object could go out of scope while it is being passed out of band.
> 
> But once the object is recreated in consumer, the original owner can go away, and third consumer can obtain the object. Or not?
> 
> So does ownership transfer in Ray if ObjectRefs are send in-band? Or is ownership shared? Or something else? (Is there documentation about this?)

No, the consumer is not guaranteed to obtain the object. It will only be guaranteed to receive the value if the object is fetched into its local plasma store and the consumer pins it there by obtaining a pointer to the object.

Ownership never transfers. We don't have good documentation on this right now, but we're working on a whitepaper that will include this.

> 
> > it's impossible to do ref counting for serialized copies of an ObjectRef.
> 
> I am not saying that. But you can count all deserialized copies of ObjectRef everywhere. Is Ray doing that?

Ray is not doing that. As I said, you're welcome to open a feature request for it to gauge interest.

> 
> > Again, not a high priority since there's a reasonable workaround and would suggest opening a separate feature request for it.
> 
> Which workaround you have in mind here? Cloudpickle? Using cloudpickle addresses this and makes ObjectRefs be counted also in consumers? Because I thought only original owners are counted?

The workaround is to use cloudpickle. The object will get pinned forever at the owner if it is serialized out of band (sorry didn't mention this before, had to confirm).

**Comment by mitar** [comment]
> We don't have good documentation on this right now, but we're working on a whitepaper that will include this.

Yea. I think this would be very welcome. :-) I am getting a bit confused here with pinning, reference counting, and owner-based reference counting, and how this all relates to keeping references to ObjectRefs in memory and how this related to keeping references to objects themselves in memory. Also, how are objects now discovered and retrieved in the cluster.

So sorry for all these questions. Not sure if this the best venue anymore, but ...

> It will only be guaranteed to receive the value if the object is fetched into its local plasma store and the consumer pins it there by obtaining a pointer to the object.

And this pinning increased the reference count?

> The object will get pinned forever at the owner if it is serialized out of band (sorry didn't mention this before, had to confirm).

So there is a side-effect here which happens when serializing using cloudpickle?

**Comment by stephanie-wang** [comment]
> And this pinning increased the reference count?

Yes, but this pinning is not related to the ref count that's kept by the owner. Confusing, I know. But the client to the plasma store keeps a separate unrelated ref count for objects in use to prevent them from being evicted while Python has a pointer to the object.

> 
> > The object will get pinned forever at the owner if it is serialized out of band (sorry didn't mention this before, had to confirm).
> 
> So there is a side-effect here which happens when serializing using cloudpickle?

Yes. [Here's](https://github.com/ray-project/ray/blob/9b1772253f47d20d8aa1bc727d67630b2026b9e5/python/ray/serialization.py#L207) the relevant line of code, if you're interested.

**Comment by mitar** [comment]
Thank you so much. What does "captured in a remote function/actor" in a comment there mean?

**Comment by stephanie-wang** [comment]
Ah that's when an ObjectID gets captured in a function/actor definition.

```python
x = ray.put(...)

@ray.remote
def capture():
  ray.get(x)
```

**Comment by mitar** [comment]
Oh, true. Nice that you capture this as well. :-)

OK, my final question (and really thank you for all the answers). If I have a task A which creates some object, calls `ray.put` on it, obtains `objectRef`, calls task B and passes the object ref as an argument, and then task A does not hold a reference to that object anymore. What happens if task B (its worker) dies and has to be restarted. Given that it was the only one holding a reference to that object, is the object still available or it was removed because owner (task A) was not holding the reference anymore?

**Comment by stephanie-wang** [comment]
> OK, my final question (and really thank you for all the answers). If I have a task A which creates some object, calls `ray.put` on it, obtains `objectRef`, calls task B and passes the object ref as an argument, and then task A does not hold a reference to that object anymore. What happens if task B (its worker) dies and has to be restarted. Given that it was the only one holding a reference to that object, is the object still available or it was removed because owner (task A) was not holding the reference anymore?

The ref counting is based on both the Python ref count of any local instances of `ObjectRef` and the submitted task count, i.e. the number of tasks that depend on the object. The submitted task count gets decremented when the task successfully finishes. For `ray.put` objects, the object gets pinned at the original creator's node until both ref counts go to 0. So in the scenario you're describing, the object would get pinned until B was successfully retried (assuming that you're using `max_retries>0`, see the [docs](https://docs.ray.io/en/latest/fault-tolerance.html)), even if A's ref has already gone out of scope.

No worries about the questions; actually these are good topics to cover in the whitepaper :)

**Comment by mitar** [comment]
Thanks!

**Comment by mitar** [comment]
I think this is waiting for a documentation improvement.

**Comment by mitar** [comment]
> Note that we don't explicitly support out-of-band object refs, so it isn't guaranteed to work well with the current ref counting protocol.

Moreover, it would be great if out-of-band object refs could be officially supported.
